hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/HazelcastWriters.java

```diff
@@ -111,62 +116,129 @@ private HazelcastWriters() {
         boolean isLocal = clientConfig == null;
         return preferLocalParallelismOne(new HazelcastWriterSupplier<UpdateMapContext<K, V, T>, T>(
                 serializableConfig(clientConfig),
-                procContext -> new UpdateMapContext<>(mapName, toKeyFn, updateFn, isLocal),
+                instance -> procContext -> new UpdateMapContext<>(instance, mapName, toKeyFn, updateFn, isLocal),
                 UpdateMapContext::add,
-                instance -> context -> context.flush(instance),
-                DistributedConsumer.noop()
+                instance -> UpdateMapContext::flush,
+                UpdateMapContext::finish
         ));
     }
 
     private static final class UpdateMapContext<K, V, T> {
-        private final String mapName;
+        private static final int MAX_PARALLEL_ASYNC_OPS = 1000;
+
         private final DistributedFunction<? super T, ? extends K> toKeyFn;
+        private final DistributedBiFunction<? super V, ? super T, ? extends V> updateFn;
         private final boolean isLocal;
-        private final ApplyFnEntryProcessor<K, V, T> entryProcessor;
+        private final PartitionService partitionService;
 
-        private IMap map;
-        private final List<T> buffer = new ArrayList<>();
-        private final Map<K, T> tmpMap = new HashMap<>();
+        private final Semaphore concurrentAsyncOpsSemaphore = new Semaphore(MAX_PARALLEL_ASYNC_OPS);
+        private final AtomicReference<Throwable> firstError = new AtomicReference<>();
+        private final IMap map;
+        private final Map<K, Object>[] tmpMaps;
 
         UpdateMapContext(
+                HazelcastInstance instance,
                 String mapName,
                 @Nonnull DistributedFunction<? super T, ? extends K> toKeyFn,
                 @Nonnull DistributedBiFunction<? super V, ? super T, ? extends V> updateFn,
                 boolean isLocal) {
-            this.mapName = mapName;
             this.toKeyFn = toKeyFn;
+            this.updateFn = updateFn;
             this.isLocal = isLocal;
-            entryProcessor = new ApplyFnEntryProcessor<>(tmpMap, updateFn);
+
+            map = instance.getMap(mapName);
+            partitionService = instance.getPartitionService();
+            int partitionCount = partitionService.getPartitions().size();
+            tmpMaps = new Map[partitionCount];
+            for (int i = 0; i < partitionCount; i++) {
+                tmpMaps[i] = new HashMap<>();
+            }
         }
 
         void add(T item) {
-            buffer.add(item);
+            K key = toKeyFn.apply(item);
+            // Calculating partitionId requires serializing - we serialize the key twice unnecessarily, but I don't
+            // know a way to avoid it.
+            int partitionId = partitionService.getPartition(key).getPartitionId();
```

Comment:
this can be a problem for the maps which has a non-default partition strategy
```
        Config config = new Config();
        MapConfig mapConfig = config.getMapConfig("map");
        PartitioningStrategyConfig partitioningStrategyConfig = new PartitioningStrategyConfig();
        partitioningStrategyConfig.setPartitioningStrategyClass(StringPartitioningStrategy.class.getName());
        mapConfig.setPartitioningStrategyConfig(partitioningStrategyConfig);

        HazelcastInstanceProxy hz = ((HazelcastInstanceProxy)Hazelcast.newHazelcastInstance(config));
        PartitionService partitionService = hz.getPartitionService();

        String key = "key@actualPartition";

        Partition partition = partitionService.getPartition(key);
        System.out.println(partition);

        Data serializedKey = hz.getSerializationService().toData(key, StringPartitioningStrategy.INSTANCE);
        Partition partitionFromSerialized = partitionService.getPartition(serializedKey);
        System.out.println(partitionFromSerialized);
```

Reply:
created an issue for it 
https://github.com/hazelcast/hazelcast/issues/13950

Reply:
fixed: when writing locally, we'll use correct partitioning strategy. When writing remotely, we can't, there's no api on client